lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2437,6 +2498,368 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
+                      [&](AccessScope typeAccessScope,
+                          const TypeRepr *complainRepr,
+                          DowngradeToWarning downgradeToWarning) {
+        auto diag = TC.diagnose(P->getLoc(),
+                                diag::pattern_type_not_usable_from_inline,
+                                anyVar->isLet(),
+                                isTypeContext);
+        highlightOffendingType(TC, diag, complainRepr);
+      });
+    });
+    return;
+  }
+
+  case DeclKind::TypeAlias: {
+    auto TAD = cast<TypeAliasDecl>(D);
+
+    checkTypeAccess(TC, TAD->getUnderlyingTypeLoc(), TAD,
+                    /*checkUsableFromInline=*/true,
+                    [&](AccessScope typeAccessScope,
+                        const TypeRepr *complainRepr,
+                        DowngradeToWarning downgradeToWarning) {
+      auto diag =
+        TC.diagnose(TAD,
+                    diag::type_alias_underlying_type_not_usable_from_inline);
+      highlightOffendingType(TC, diag, complainRepr);
+    });
+
+    return;
+  }
+
+  case DeclKind::AssociatedType: {
+    auto assocType = cast<AssociatedTypeDecl>(D);
+
+    // This must stay in sync with diag::associated_type_not_usable_from_inline.
+    enum {
+      ACEK_DefaultDefinition = 0,
+      ACEK_Requirement
+    };
+
+    std::for_each(assocType->getInherited().begin(),
+                  assocType->getInherited().end(),
+                  [&](TypeLoc requirement) {
+      checkTypeAccess(TC, requirement, assocType,
+                      /*checkUsableFromInline=*/true,
+                      [&](AccessScope typeAccessScope,
+            const TypeRepr *complainRepr,
+            DowngradeToWarning downgradeDiag) {
+        auto diag = TC.diagnose(assocType,
+                                diag::associated_type_not_usable_from_inline,
+                                ACEK_Requirement);
+        highlightOffendingType(TC, diag, complainRepr);
+      });
+    });
+    checkTypeAccess(TC, assocType->getDefaultDefinitionLoc(), assocType,
+                    /*checkUsableFromInline=*/true,
+                    [&](AccessScope typeAccessScope,
+                        const TypeRepr *complainRepr,
+                        DowngradeToWarning downgradeDiag) {
+      auto diag = TC.diagnose(assocType,
+                              diag::associated_type_not_usable_from_inline,
+                              ACEK_DefaultDefinition);
+      highlightOffendingType(TC, diag, complainRepr);
+    });
+
+    return;
+  }
+
+  case DeclKind::Enum: {
+    auto ED = cast<EnumDecl>(D);
+
+    checkGenericParamAccess(TC, ED->getGenericParams(), ED,
+                            /*checkUsableFromInline=*/true);
+
+    if (ED->hasRawType()) {
+      Type rawType = ED->getRawType();
+      auto rawTypeLocIter = std::find_if(ED->getInherited().begin(),
+                                         ED->getInherited().end(),
+                                         [&](TypeLoc inherited) {
```

Comment:
Okay this is really starting to seem like too much code duplicated from the regular access checking.

Reply:
Yeah. Perhaps we can factor out a visitor that walks the TypeLocs in a Decl?